crates/consensus/debug-client/src/client.rs

```diff
@@ -1,84 +1,60 @@
+}
+
+/// Debyg consensus client that sends FCUs and new payloads using recent blocks from an external
+/// provider like Etherscan or an RPC endpoint.
 #[derive(Debug)]
-pub struct RpcConsensusClient {
-    /// HTTP client to fetch blocks
-    http_client: reqwest::Client,
-    /// Handle to execution client
+pub struct DebugConsensusClient<P: BlockProvider> {
+    /// Handle to execution client.
     auth_server: AuthServerHandle,
-    /// Etherscan API key
-    etherscan_api_key: String,
-    /// Etherscan base API URL
-    etherscan_base_api_url: String,
+    /// Provider to get consensus blocks from.
+    block_provider: Arc<P>,
 }
 
-impl RpcConsensusClient {
-    /// Create a new fake consensus client that should sent FCUs and new payloads to `auth_server`.
-    pub fn new(
-        auth_server: AuthServerHandle,
-        etherscan_api_key: String,
-        etherscan_base_api_url: String,
-    ) -> Self {
-        Self {
-            http_client: reqwest::Client::new(),
-            auth_server,
-            etherscan_api_key,
-            etherscan_base_api_url,
-        }
+impl<P: BlockProvider> DebugConsensusClient<P> {
+    /// Create a new debug consensus client with the given handle to execution
+    /// client and block provider.
+    pub fn new(auth_server: AuthServerHandle, block_provider: Arc<P>) -> Self {
+        Self { auth_server, block_provider }
     }
 
-    /// Spawn the client to start sending FCUs and new payloads by periodically fetching recent blocks.
+    /// Spawn the client to start sending FCUs and new payloads by periodically fetching recent
+    /// blocks.
     pub async fn spawn<T: EngineTypes>(&self) {
         // TODO: Add logs
-        // TODO: Generalize over block-fetching code to support different sources
-
         let execution_client = self.auth_server.http_client();
-        let mut last_block_number: Option<u64> = None;
-
-        loop {
-            let block: EtherscanBlockResponse = self
-                .http_client
-                .get(&self.etherscan_base_api_url)
-                .query(&[
-                    ("module", "proxy"),
-                    ("action", "eth_getBlockByNumber"),
-                    ("tag", "latest"),
-                    ("boolean", "true"),
-                    ("apikey", &self.etherscan_api_key),
-                ])
-                .send()
-                .await
-                .unwrap()
-                .json()
-                .await
-                // TODO: Handle errors gracefully and do not stop the loop
-                .unwrap();
-
-            // Sleep if no new block is available
-            if block.result.header.number == last_block_number {
-                // TODO: Allow to configure this
-                sleep(Duration::from_secs(3)).await;
-                continue;
-            }
-
-            let payload = rich_block_to_execution_payload_v3(block.result);
+        let mut previous_block_hashes = AllocRingBuffer::new(64);
```

Comment:
why do we need this?

Reply:
we're using head-64, head-32 as finalized and safe blocks. `previous_block_hashes` stores last 64 blockhashes for this purpose.